// @antv/l7-utils/lib/ajax.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.postData = exports.getTiffImage = exports.getJSON = exports.getImage = exports.getArrayBuffer = exports.AJAXError = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _wrapNativeSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/wrapNativeSuper"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _env = require("./env");

var _miniAdapter = require("./mini-adapter");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var AJAXError = function (_Error) {
  (0, _inherits2.default)(AJAXError, _Error);

  var _super = _createSuper(AJAXError);

  function AJAXError(status, statusText, url, body) {
    var _this;

    (0, _classCallCheck2.default)(this, AJAXError);
    _this = _super.call(this, "AJAXError: ".concat(statusText, " (").concat(status, "): ").concat(url));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "status", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "statusText", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "url", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "body", void 0);
    _this.status = status;
    _this.statusText = statusText;
    _this.url = url;
    _this.body = body;
    return _this;
  }

  return (0, _createClass2.default)(AJAXError);
}((0, _wrapNativeSuper2.default)(Error));

exports.AJAXError = AJAXError;

function makeFetchRequest(requestParameters, callback) {
  var request = new Request(requestParameters.url, {
    method: requestParameters.method || 'GET',
    body: requestParameters.body,
    credentials: requestParameters.credentials,
    headers: requestParameters.headers,
    referrer: (0, _env.getReferrer)(),
    signal: requestParameters === null || requestParameters === void 0 ? void 0 : requestParameters.signal
  });

  if (requestParameters.type === 'json') {
    request.headers.set('Accept', 'application/json');
  }

  return fetch(request).then(function (response) {
    if (response.ok) {
      return (requestParameters.type === 'arrayBuffer' ? response.arrayBuffer() : requestParameters.type === 'json' ? response.json() : response.text()).then(function (result) {
        callback(null, result, response.headers.get('Cache-Control'), response.headers.get('Expires'));
      }).catch(function (err) {
        callback(new Error(err.message));
      });
    }

    return response.blob().then(function (body) {
      return callback(new AJAXError(response.status, response.statusText, requestParameters.url, body));
    });
  }).catch(function (error) {
    callback(new Error(error.message));
  });
}

function makeXMLHttpRequest(requestParameters, callback) {
  var xhr = new _miniAdapter.$XMLHttpRequest();
  xhr.open(requestParameters.method || 'GET', requestParameters.url, true);

  if (requestParameters.type === 'arrayBuffer') {
    xhr.responseType = 'arraybuffer';
  }

  for (var k in requestParameters.headers) {
    if (requestParameters.headers.hasOwnProperty(k)) {
      xhr.setRequestHeader(k, requestParameters.headers[k]);
    }
  }

  if (requestParameters.type === 'json') {
    xhr.responseType = 'text';
    xhr.setRequestHeader('Accept', 'application/json');
  }

  xhr.withCredentials = requestParameters.credentials === 'include';

  xhr.onerror = function () {
    callback(new Error(xhr.statusText));
  };

  xhr.onload = function () {
    if ((xhr.status >= 200 && xhr.status < 300 || xhr.status === 0) && xhr.response !== null) {
      var _data = xhr.response;

      if (requestParameters.type === 'json') {
        try {
          _data = JSON.parse(xhr.response);
        } catch (err) {
          return callback(err);
        }
      }

      callback(null, _data, xhr.getResponseHeader('Cache-Control'), xhr.getResponseHeader('Expires'));
    } else {
      var body = new Blob([xhr.response], {
        type: xhr.getResponseHeader('Content-Type')
      });
      callback(new AJAXError(xhr.status, xhr.statusText, requestParameters.url, body));
    }
  };

  xhr.send(requestParameters.body);
  return xhr;
}

function makeRequest(requestParameters, callback) {
  return makeXMLHttpRequest(requestParameters, callback);
}

var getJSON = function getJSON(requestParameters, callback) {
  return makeRequest(_objectSpread(_objectSpread({}, requestParameters), {}, {
    type: 'json'
  }), callback);
};

exports.getJSON = getJSON;

var getArrayBuffer = function getArrayBuffer(requestParameters, callback) {
  return makeRequest(_objectSpread(_objectSpread({}, requestParameters), {}, {
    type: 'arrayBuffer'
  }), callback);
};

exports.getArrayBuffer = getArrayBuffer;

var postData = function postData(requestParameters, callback) {
  return makeRequest(_objectSpread(_objectSpread({}, requestParameters), {}, {
    method: 'POST'
  }), callback);
};

exports.postData = postData;

function sameOrigin(url) {
  var a = _miniAdapter.$window.document.createElement('a');

  a.href = url;
  return a.protocol === _miniAdapter.$window.document.location.protocol && a.host === _miniAdapter.$window.document.location.host;
}

var transparentPngUrl = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAC0lEQVQYV2NgAAIAAAUAAarVyFEAAAAASUVORK5CYII=';

function arrayBufferToImage(data, callback) {
  var img = new _miniAdapter.$window.Image();
  var URL = _miniAdapter.$window.URL || _miniAdapter.$window.webkitURL;
  img.crossOrigin = 'anonymous';

  img.onload = function () {
    callback(null, img);
    URL.revokeObjectURL(img.src);
    img.onload = null;
    window.requestAnimationFrame(function () {
      img.src = transparentPngUrl;
    });
  };

  img.onerror = function () {
    return callback(new Error('Could not load image. Please make sure to use a supported image type such as PNG or JPEG. Note that SVGs are not supported.'));
  };

  var blob = new Blob([new Uint8Array(data)], {
    type: 'image/png'
  });
  img.src = data.byteLength ? URL.createObjectURL(blob) : transparentPngUrl;
}

function arrayBufferToImageBitmap(data, callback) {
  var blob = new Blob([new Uint8Array(data)], {
    type: 'image/png'
  });
  createImageBitmap(blob).then(function (imgBitmap) {
    callback(null, imgBitmap);
  }).catch(function (e) {
    callback(new Error("Could not load image because of ".concat(e.message, ". Please make sure to use a supported image type such as PNG or JPEG. Note that SVGs are not supported.")));
  });
}

var getImage = function getImage(requestParameters, callback) {
  return getArrayBuffer(requestParameters, function (err, imgData) {
    if (err) {
      callback(err);
    } else if (imgData) {
      var imageBitmapSupported = typeof createImageBitmap === 'function';

      if (imageBitmapSupported) {
        arrayBufferToImageBitmap(imgData, callback);
      } else {
        arrayBufferToImage(imgData, callback);
      }
    }
  });
};

exports.getImage = getImage;

var arrayBufferToTiffImage = function () {
  var _ref = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee(data, callback, rasterParser) {
    var _yield$rasterParser, rasterData, width, height, defaultMIN, defaultMAX;

    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.prev = 0;
            _context.next = 3;
            return rasterParser(data);

          case 3:
            _yield$rasterParser = _context.sent;
            rasterData = _yield$rasterParser.rasterData;
            width = _yield$rasterParser.width;
            height = _yield$rasterParser.height;
            defaultMIN = 0;
            defaultMAX = 8000;
            callback(null, {
              data: rasterData,
              width: width,
              height: height,
              min: defaultMIN,
              max: defaultMAX
            });
            _context.next = 15;
            break;

          case 12:
            _context.prev = 12;
            _context.t0 = _context["catch"](0);
            callback(null, new Error('' + _context.t0));

          case 15:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, null, [[0, 12]]);
  }));

  return function arrayBufferToTiffImage(_x, _x2, _x3) {
    return _ref.apply(this, arguments);
  };
}();

var getTiffImage = function getTiffImage(requestParameters, callback, rasterParser) {
  return getArrayBuffer(requestParameters, function (err, imgData) {
    if (err) {
      callback(err);
    } else if (imgData) {
      arrayBufferToTiffImage(imgData, callback, rasterParser);
    }
  });
};

exports.getTiffImage = getTiffImage;
