// @ant-design/pro-form/lib/components/SchemaForm/valueType/formList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.formList = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _jsxRuntime = require("react/jsx-runtime");

var _List = require("../../List");

var formList = function formList(item, _ref) {
  var genItems = _ref.genItems;

  if (item.valueType === 'formList' && item.dataIndex) {
    var _item$getFieldProps;

    if (!item.columns || !Array.isArray(item.columns)) return null;
    return (0, _jsxRuntime.jsx)(_List.ProFormList, (0, _objectSpread2.default)((0, _objectSpread2.default)({
      name: item.dataIndex,
      label: item.label,
      initialValue: item.initialValue,
      colProps: item.colProps,
      rowProps: item.rowProps
    }, (_item$getFieldProps = item.getFieldProps) === null || _item$getFieldProps === void 0 ? void 0 : _item$getFieldProps.call(item)), {}, {
      children: genItems(item.columns)
    }), item.key);
  }

  return true;
};

exports.formList = formList;
